pkg/tiup/deploy.go

```diff
@@ -579,3 +691,158 @@ func (c *ClusterManager) GenUnzipCommand(filePath string, destDir string) string
+			if reflect.DeepEqual(pre.PD[i], cur.PD[j]) {
+				isExisted = true
+			}
+		}
+		if !isExisted {
+			if pre.PD[i].ClientPort == 0 {
+				pre.PD[i].ClientPort = 2379
+			}
+			node := clusterIPMaps[pre.PD[i].Host] + ":" + strconv.Itoa(pre.PD[i].ClientPort)
+			result = append(result, node)
+		}
+	}
+	for i := 0; i < len(pre.TiKV); i++ {
+		var isExisted bool
+		for j := 0; j < len(cur.TiKV); j++ {
+			if reflect.DeepEqual(pre.TiKV[i], cur.TiKV[j]) {
+				isExisted = true
+			}
+		}
+		if !isExisted {
+			if pre.TiKV[i].Port == 0 {
+				pre.TiKV[i].Port = 20160
+			}
+			node := clusterIPMaps[pre.TiKV[i].Host] + ":" + strconv.Itoa(pre.TiKV[i].Port)
+			result = append(result, node)
+		}
+	}
+	return result
+}
+
+func (c *ClusterManager) getScaleOutRoles(pre naglfarv1.TiDBCluster, cur naglfarv1.TiDBCluster, clusterIPMaps map[string]string) ([]naglfarv1.TiDBSpec, []naglfarv1.PDSpec, []naglfarv1.TiKVSpec) {
+
+	var tidbs []naglfarv1.TiDBSpec
+	for i := 0; i < len(cur.TiDB); i++ {
+		var isExisted bool
+		for j := 0; j < len(pre.TiDB); j++ {
+			if reflect.DeepEqual(cur.TiDB[i], pre.TiDB[j]) {
+				isExisted = true
+			}
+		}
+		if !isExisted {
+			tmp := cur.TiDB[i]
+			tmp.Host = clusterIPMaps[tmp.Host]
+			// if nil, set default
+			if tmp.SshPort == 0 {
+				tmp.SshPort = 22
+			}
+			if tmp.Port == 0 {
+				tmp.Port = 4000
+			}
+			if tmp.StatusPort == 0 {
+				tmp.StatusPort = 10080
+			}
+			if tmp.DeployDir == "" {
+				tmp.DeployDir = "/data/deploy/install/deploy/tidb-4000"
+			}
+			if tmp.LogDir == "" {
+				tmp.LogDir = "/data/deploy/install/log/tidb-4000"
+			}
+			tidbs = append(tidbs, tmp)
+		}
+	}
+
+	var pds []naglfarv1.PDSpec
+	for i := 0; i < len(cur.PD); i++ {
+		var isExisted bool
+		for j := 0; j < len(pre.PD); j++ {
+			if reflect.DeepEqual(cur.PD[i], pre.PD[j]) {
+				isExisted = true
+			}
+		}
+		if !isExisted {
+			tmp := cur.PD[i]
+			tmp.Host = clusterIPMaps[tmp.Host]
+
+			if tmp.SshPort == 0 {
+				tmp.SshPort = 22
+			}
+			if tmp.ClientPort == 0 {
+				tmp.ClientPort = 2379
+			}
+			if tmp.PeerPort == 0 {
+				tmp.PeerPort = 2380
+			}
+			if tmp.DataDir == "" {
+				tmp.DataDir = "/data/deploy/install/data/pd-2379"
+			}
+			if tmp.DeployDir == "" {
+				tmp.DeployDir = "/data/deploy/install/deploy/pd-2379"
+			}
+			if tmp.LogDir == "" {
+				tmp.LogDir = "/data/deploy/install/log/pd-2379"
+			}
+
+			pds = append(pds, tmp)
+		}
+	}
+
+	var tikvs []naglfarv1.TiKVSpec
+	for i := 0; i < len(cur.TiKV); i++ {
+		var isExisted bool
+		for j := 0; j < len(pre.TiKV); j++ {
+			if reflect.DeepEqual(cur.TiKV[i], pre.TiKV[j]) {
+				isExisted = true
+			}
+		}
+		if !isExisted {
+			tmp := cur.TiKV[i]
+			tmp.Host = clusterIPMaps[tmp.Host]
+
+			if tmp.SshPort == 0 {
+				tmp.SshPort = 22
+			}
+			if tmp.Port == 0 {
+				tmp.Port = 20160
+			}
+			if tmp.StatusPort == 0 {
+				tmp.StatusPort = 20180
+			}
+			if tmp.DataDir == "" {
+				tmp.DataDir = "/data/deploy/install/data/tikv-20160"
```

Comment:
If I mount a nvme device on /disk1...

Reply:
This needs to be discussed. I didn't know the use of nvme before.